### ISSUE TITLE
More use of opportunistic dupe suppression

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#5055]
1. The current code implemented in #5046 does not suppress dupes produced by `--rules-stack` (but *does* suppress dupes right before them being applied). Example:
```
$ rm -f out && ../run/john -loopback -stdout -rules=single -rules-stack=best64 | unique out
Using default input encoding: UTF-8
Permutation rules: single
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
Enabling duplicate candidate password suppressor
268275585p 0:00:01:36 100.00% (2022-02-07 14:29) 2789Kp/s A21900
Total lines read: 268275585, unique lines written: 146550958 (54%), 1 slow passes
```

2. We might want to enable use of dupe suppression for `--external` mode, as some of them can produce many dupes.
```
$ rm -f out && ../run/john -stdout -external:subsets -max-len=6 | unique out
Warning: external mode 'subsets' can't be resumed if aborted
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
150042430p 0:00:00:45 100.00% (2022-02-07 14:33) 3315Kp/s ~~~~~~
Total lines read: 150042430, unique lines written: 101829360 (67%), no slow passes
```

3. We could consider `--regex` mode. While most carefully thought out regex'es produce no dupes, I think there are situations where they are hard to battle. And there can definitely be many dupes from hybrid wordlist + regex.

**Comment by solardiz** [comment]
> 1. does not suppress dupes produced by `--rules-stack` (but _does_ suppress dupes right before them being applied)

Ideally we'd have both, and here's an idea how to accomplish that without having two separate filters (and have to split the memory usage between them): in one of the two invocations, we can apply a tweak to the hash function such that we would be very unlikely to match on a wrong-level candidate (pre vs. post stacked rules). Unfortunately, to make this in fact "very unlikely" we'll have to give up on the nice guaranteed-non-collision properties for up to length 8 in one of those two invocations. Also, if the tweak has performance cost, we'll prefer to apply it to the pre stacked rules invocations (since these are less numerous). An easy hack achieving these properties would be to prefix the pre stacked rules non-final candidate password strings with some weird 8-byte string for the purpose of dupe checker (only), or optimizing that to encode (the effect of) the same into the hash function when a flag is set (no reason to go over the fixed initial 8 bytes over and over again, we can instead reuse the state after that point).

**Comment by solardiz** [comment]
> we'll have to give up on the nice guaranteed-non-collision properties for up to length 8 in one of those two invocations.

However, perhaps it's still possible to have similar for up to length 7, by encoding the difference between two invocations in the 8th byte.

**Comment by solardiz** [comment]
As the opposite case to stacked rules, when running rules and mask (e.g., `-rules -mask='?w?d'`) and not having on-device mask, the suppressor is currently applied to final post-mask candidates. I think we could improve efficiency (both speed and percentage suppressed by a certain size) by applying it before mask instead, because the mask cannot produce duplicates (right?) if the pre-mask stream is all unique.

**Comment by solardiz** [comment]
> tweak to the hash function such that we would be very unlikely to match on a wrong-level candidate (pre vs. post stacked rules).

As an option, we could simply - and externally to the suppressor - prepend or append a linefeed character to the password in one of the two invocations of the suppressor. This character currently can't come from a wordlist, and rules are unlikely (aren't supposed to) introduce it.

**Comment by solardiz** [comment]
> Ideally we'd have both, and here's an idea how to accomplish that without having two separate filters

A weird aspect of using a filter twice is that the statistics would be of a mix of both uses. Even the logic to auto-disable the filter would be based on its cumulative efficiency and p/s rate at both levels. This isn't necessarily wrong, but I guess in some cases it'd be best for the filter to only auto-disable itself at the post-stacked-rules level (where its overhead is higher) and stay enabled at pre-stacked-rules (until possibly reaching a separate threshold). Otherwise things could sometimes be worse than they're now (the filter would unnecessarily get disabled pre-stacked-rules as well, whereas now it could stay enabled because the p/s rate at that level is lower).

**Comment by magnumripper** [comment]
BTW when running fork or MPI, multiple messages are printed for enabling/disabling the suppressor. We might want the "Enabling..." to be printed only for `john_main_process` and have the "Disabling..." printed with node number when applicable.

Edit: #5062

**Comment by solardiz** [comment]
> We might want to enable use of dupe suppression for `--external` mode, as some of them can produce many dupes.

Yes, and also when external filters are used on top of another mode, which is now relevant for `--external=Untokenize` as produced by `tokenize.pl`. For this, we'd probably need to allocate and use an extra options flag separate from `FLG_RULES_ALLOW`, e.g. called `FLG_DUPESUPP_ALLOW`.
